File: baseble/src/main/java/com/fcbox/ble/baseble/core/LruHashMap.java
package com.fcbox.ble.baseble.core;

import java.util.LinkedHashMap;


/**
 *  Lru算法实现的HashMap
 * @param <K> 键
 * @param <V> 值
 */
public class LruHashMap<K, V> extends LinkedHashMap<K, V> {
    private final int MAX_SAVE_SIZE;

    public LruHashMap(int saveSize) {
        super((int) Math.ceil(saveSize / 0.75) + 1, 0.75f, true);
        MAX_SAVE_SIZE = saveSize;
    }

    @Override
    protected boolean removeEldestEntry(Entry eldest) {
        if (size() > MAX_SAVE_SIZE && eldest.getValue() instanceof DeviceMirror) {
            ((DeviceMirror) eldest.getValue()).disconnect();
        }
        return size() > MAX_SAVE_SIZE;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Entry<K, V> entry : entrySet()) {
            sb.append(String.format("%s:%s ", entry.getKey(), entry.getValue()));
        }
        return sb.toString();
    }
}

File: baseble/src/main/java/com/xzy/ble/baseble/core/LruHashMap.kt
package com.xzy.ble.baseble.core

import java.util.LinkedHashMap


/**
 * Lru算法实现的 HashMap
 * @param <K> 键
 * @param <V> 值
</V></K> */
class LruHashMap<K, V>(private val MAX_SAVE_SIZE: Int) : LinkedHashMap<K, V>(Math.ceil(MAX_SAVE_SIZE / 0.75).toInt() + 1, 0.75f, true) {

    override fun removeEldestEntry(eldest: MutableMap.MutableEntry<K, V>?): Boolean {
        if (eldest != null) {
            if (size > MAX_SAVE_SIZE && eldest.value is DeviceMirror) {
                (eldest.value as DeviceMirror).disconnect()
            }
        }
        return size > MAX_SAVE_SIZE
    }

    override fun toString(): String {
        val sb = StringBuilder()
        for ((key, value) in entries) {
            sb.append(String.format("%s:%s ", key, value))
        }
        return sb.toString()
    }
}
